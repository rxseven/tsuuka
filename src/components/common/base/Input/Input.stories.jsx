import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import currencySchema from 'schemas/currency';
import Input from './index';

const title = 'Components/Common/Base/Input';
const props = { onChange: action('onChange'), ref: React.createRef() };
const validator = value => {
  if (value && value !== 'Oops!') {
    return currencySchema.isValidSync({ amount: value });
  }

  return true;
};

storiesOf(title, module)
  .addParameters({
    info: {
      text: `
  **Usage:** for more information, see [Bootstrap - Forms](https://getbootstrap.com/docs/4.1/components/forms/).
`
    }
  })
  .add('text input (default)', () => <Input {...props} />, {
    notes: {
      markdown: `As this component implements **React.forwardRef**, the story source will not show the correct data source. Furthermore, DevTool will also show a warning message addressing an invalid prop type.`
    }
  })
  .add('password', () => <Input {...props} type="password" />)
  .add('disabled', () => <Input {...props} disabled />)
  .add('with placeholder', () => <Input {...props} placeholder="Text input" />)
  .add('with value', () => <Input {...props} value="username@mail.com" />)
  .add('with validator', () => (
    <Input {...props} validator={validator} value="Invalid value" />
  ));
