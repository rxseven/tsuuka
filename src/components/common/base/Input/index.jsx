import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

const propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  type: PropTypes.string,
  validator: PropTypes.oneOfType([PropTypes.func, PropTypes.any]),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

const defaultProps = {
  className: '',
  type: 'text',
  validator: null,
  value: ''
};

const Input = React.forwardRef(
  ({ className, validator, value, ...props }, ref) => {
    return (
      <input
        {...props}
        className={cx(
          'form-control',
          validator && (!validator(value) && 'is-invalid'),
          className
        )}
        ref={ref}
        value={value}
      />
    );
  }
);

Input.propTypes = propTypes;
Input.defaultProps = defaultProps;

export default Input;
