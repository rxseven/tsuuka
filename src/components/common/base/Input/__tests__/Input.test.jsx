import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Input from '../index';

describe('<Input />', () => {
  // Arrange
  const props = { onChange: jest.fn() };
  const component = <Input {...props} />;

  describe('Unit tests', () => {
    it('should render without crashing', () => {
      // Act
      const wrapper = shallow(component);

      // Assert
      expect(wrapper).toBeDefined();
    });
  });

  describe('Snapshot tests', () => {
    it('should render correctly', () => {
      // Act
      const tree = renderer.create(component).toJSON();

      // Assert
      expect(tree).toMatchSnapshot();
    });
  });
});
