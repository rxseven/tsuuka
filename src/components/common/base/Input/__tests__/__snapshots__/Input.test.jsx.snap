// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Snapshot tests should render correctly 1`] = `
<input
  className="form-control"
  onChange={[MockFunction]}
  type="text"
  value=""
/>
`;
